Fixed #24943 -- Update contributing tutorial to use virtualenv

alimony opened the pull request:
See ticket: https://code.djangoproject.com/ticket/24943

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -68,6 +68,16 @@ It contains lots of great information and is a must read for anyone who'd like
 
+.. note::
+
+    All of the instructions in this tutorial assumes you are using Python 3. You
```
This tutorial assumes...

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -68,6 +68,16 @@ It contains lots of great information and is a must read for anyone who'd like
+
+    All of the instructions in this tutorial assumes you are using Python 3. You
+    might want to install it if you don't have it already. The easiest way is
+    through a package manager, by running e.g. ``apt-get install python3`` or
```
We could borrow the language we use on topics/install: "Get the latest version of Python at https://www.python.org/download/ or with your operating system’s package manager."

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +113,84 @@ Download the Django source code repository using the following command:
+install any package using ``pip``. The most convenient way to do so is by using
+a *virtual environment* (or virtualenv) which is a feature built into Python
+that allows you to keep a separate directory of installed packages for each of
+your projects so that these don't interfere with each other.
```
so that they

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +113,84 @@ Download the Django source code repository using the following command:
+.. code-block:: none
+
+    > mkdir C:\Users\Name\virtualenvs
+    > C:\Python34\python -m venv C:\Users\Name\virtualenvs\djangodev
```
Let's check if python is added to the path on a default install so we don't need to reference C:\Python34

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +113,84 @@ Download the Django source code repository using the following command:
+
+.. code-block:: none
+
+    > C:\Users\Name\virtualenvs\djangodev\Scripts\activate
```
does ~ expansion not work on Windows?

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +113,84 @@ Download the Django source code repository using the following command:
+
+    > C:\Users\Name\virtualenvs\djangodev\Scripts\activate
+
+Anything you install through ``pip`` from now on will be installed in your new
```
Might be helpful to mention that you need to active the virtualenv each time you open a new shell.

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -164,27 +232,20 @@ into the Django ``tests/`` directory and then running:
-    $ PYTHONPATH=.. ./runtests.py
+    $ ./runtests.py
 
 If you're on Windows, the above should work provided that you are using
```
Maybe it would be worth introducing Git Bash from the start as that might help eliminate Windows differences.

alimony commented:
Updated according to review. I went through the tutorial on Windows and noted the following:
* A fresh install of Python 3.4 **does not** seem to add it to default path.
* If the "Git Bash" option is selected on installing Git, everything else will be smoother, so moved that recommendation to the top.
* The built-in venv functionality of Python 3.4 on Windows seems only to install `activate.bat` and `Activate.ps1` scripts, i.e. nothing suitable for the Git Bash shell. Installing the `virtualenv` package instead solves this, so recommending that in the tutorial.
* Windows does indeed do the tilde expansion, so updated that.
* Added a note about activating venv on new shell, but also included a link to `virtualenvwrapper` since a lot of people find it really convenient.

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -68,6 +68,13 @@ It contains lots of great information and is a must read for anyone who'd like
 
+.. note::
+
+    This tutorial assumes you are using Python 3. You might want to install it
```
chop sentence "You might want to install it..." (seems to imply it's optional)?

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -82,8 +89,14 @@ development version of Django and to generate patch files for the changes you
+line. If you get messages saying that this command could not be found, you'll
+have to download and install it, see `Git's download page`__.
+
+.. note::
```
`.. admonition:: For Windows users` will give a more helpful title in the rendered docs.

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -68,6 +68,13 @@ It contains lots of great information and is a must read for anyone who'd like
+
+    This tutorial assumes you are using Python 3. You might want to install it
+    if you don't have it already. Get the latest version of Python at
+    https://www.python.org/download/ or with your operating system's package
```
We can use text links rather than a raw URL to match the rest of this doc.
```
`Python's download page <https://www.python.org/download/>`_
```

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +116,101 @@ Download the Django source code repository using the following command:
+
+.. code-block:: none
+
+    > mkdir C:Users/name/virtualenvs
```
It's a bit odd to see no slash after `C:` - I guess it works since you tested it, but I'm curious about it.

alimony replied:
Yes it is a bit odd. I will set everything up again from scratch on a new machine and type every single line verbatim (without e.g. tab completion) to make sure these instructions work 100%.

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +116,101 @@ Download the Django source code repository using the following command:
+        $ pip3 install virtualenv
+        $ virtualenv --python=`which python3` ~/.virtualenvs/djangodev
+
+    Or on Windows:
```
As far as I know, this issue only affects Ubuntu (https://bugs.launchpad.net/ubuntu/+source/python3.4/+bug/1290847) so Windows/Mac aren't affected.  So "On older versions of Python 3" should be replaced with "On Ubuntu" I think.

alimony replied:
Actually it's needed on Windows too, see the third bullet point in my last comment:

* The built-in venv functionality of Python 3.4 on Windows seems only to install activate.bat and Activate.ps1 scripts, i.e. nothing suitable for the Git Bash shell. Installing the virtualenv package instead solves this, so recommending that in the tutorial.

I'm walking through the setup again on Windows now, but I think this will remain the case.

timgraham replied:
`souce env/Scripts/activate` works for me in Git Bash.

alimony replied:
Are you sure you have not install/setup that virtualenv using the external Python package? On an entirely new machine with a fresh install of Python 3.4 and Git, I get this:

![skarmavbild 2015-06-15 kl 19 34 51](https://cloud.githubusercontent.com/assets/331091/8166070/c69a24fe-1395-11e5-9d34-c412d1919a9f.png)

timgraham replied:
I don't remember. Fine with keeping it in then.

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +116,101 @@ Download the Django source code repository using the following command:
     .. code-block:: console
 
-        $ pip install -e /path/to/your/local/clone/django/
+        $ pip3 install virtualenv
```
I don't think pip3 works without first `apt-get install python3-pip`

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +116,101 @@ Download the Django source code repository using the following command:
+    .. code-block:: none
 
-__ http://www.virtualenv.org
+        $ C:Python34/Scripts/pip install virtualenv
```
I think it'd be worth telling Windows user to check the options to add these (and python itself) to their path during install.

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +116,101 @@ Download the Django source code repository using the following command:
+If the ``activate`` script is not found, you need to install ``virtualenv`` as
+a package, as described above.
+
+.. note::
```
use admonition as suggested earlier?

alimony commented:
Updated per all review comments, and from running the entire setup again from scratch on Windows. As suggested, managed to get rid of all ugly path components (`C:`, `Python34` and such)

timgraham commented on docs/intro/contributing.txt:
```diff
@@ -103,26 +120,96 @@ Download the Django source code repository using the following command:
 
-        $ pip install -e /path/to/your/local/clone/django/
+        $ sudo apt-get install python3-pip
+        $ pip3 install virtualenv
```
I'm a bit surprised sudo isn't required (it didn't seem to be required on 15.04 as packages are installed in ~/.local), did you test on a fresh 14.04 install). Maybe just add a comment like  `# prefix this command with sudo if it gives a permission denied error`

alimony replied:
Done!

timgraham commented:
merged in fb1ba4d63e76c418c2dd6214dabe4dcf25ab3bad, thanks!
